### PR TITLE
Add resource and action to failure, use generic whitelists to remove whitelisted failures

### PR DESCRIPTION
**Pull request by IgnacioRV**
## Context
Following the update to having a generic whitelist for resources and another one for actions, this update adds the option to add a list of resource_ids and/or actions to the failure.
It also introduces the concept of failure granularity, each failure has a granularity which can be:
- Stack: Applies to all stack
- Resource: Applies to a set of resources in the stack
- Action: Applies to a set of actions in the stack
The new methods added in the rule processor will use the granularity of the failure along with the new whitelists that were already added to filter the failures that contain affect ONLY whitelisted resources or failures according to the failure granularity.

## Changes
- Added a granularity enum
- Added methods to remove whitelisted failures and resources to the rule processor, the methods are called after processing the rules.
- Added tests to cover every possible use case

I will wait until the current PR's are merged to merge the changes into this branch and solve the possible merge conflicts here.

### CODE REVIEW

File: cfripper/model/rule.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+from cfripper.config.config import Config
+from cfripper.model.enums import RuleMode, RuleRisk, RuleGranularity
+from cfripper.model.result import Result
+
+
+class Rule(ABC):
+    RULE_MODE = RuleMode.BLOCKING
+    RISK_VALUE = RuleRisk.MEDIUM
+    GRANULATIRY = RuleGranularity.STACK
```
Comment (jsoucheiron) [inline cfripper/model/rule.py]:
```suggestion
    GRANULARITY = RuleGranularity.STACK
```

Comment (jsoucheiron) [reply cfripper/model/rule.py]: Typo :P

File: cfripper/model/rule_processor.py [hunk 2]
```diff
@@ -79,6 +56,72 @@ def process_cf_template(self, cf_template_dict, config, result):
...
+
+            whitelisted_resources = {
+                resource
+                for resource in failure["resource_ids"]
+                if any([
+                    re.match(whitelisted_resource_regex, resource)
+                    for whitelisted_resource_regex in config.get_whitelisted_resources(failure["rule"])
+                ])
+            }
+            print(f"XXXXXX {whitelisted_resources}")
```
Comment (jsoucheiron) [inline cfripper/model/rule_processor.py]: leftover print?

File: tests/test_rule_processor.py [hunk 3]
```diff
@@ -77,40 +81,453 @@
...
-    def test_with_no_rules(self):
-        processor = RuleProcessor()
-        config = Mock()
-        result = Mock()
+def test_with_no_rules():
+    processor = RuleProcessor()
+    config = Mock()
+    result = Result()
+
+    processor.process_cf_template(EXAMPLE_CF_TEMPLATE, config, result)
```
Comment (jsoucheiron) [inline tests/test_rule_processor.py]: Shouldn't there be an assert here?

File: cfripper/model/rule_processor.py (hunk 2 above)
Comment (IgnacioRV) [reply cfripper/model/rule_processor.py]: Ooops, totally

File: tests/test_rule_processor.py (hunk 3 above)
Comment (IgnacioRV) [reply tests/test_rule_processor.py]: There was none before, but I'd say yes

**Pull request merged**